### ISSUE TITLE
using ImageBufAlgo to render text?

### ISSUE THREAD

**Issue by diuming** [OpenImageIO/oiio#3943]
OS: macOS 13.4.1 Apple M1
OpenImageIO: 2.4.14.0

Am I doing something wrong?

C++ code
```
    const int WIDTH = 640, HEIGHT = 480, CHANNELS = 3;
    unsigned char buffer[WIDTH * HEIGHT * CHANNELS];
    ImageBuf ib = ImageBuf(ImageSpec(WIDTH, HEIGHT, CHANNELS), buffer);
    ImageBufAlgo::render_text(ib, 0, 10, "Hello, world!");
    ib.write("oiio-render_text.png");
```

result `oiio-render_text.png`
![oiio-render_text](https://github.com/OpenImageIO/oiio/assets/39903550/bcc9c462-3152-4d77-89f6-70ead4e6110f)

**Comment by diuming** [comment]
@apetrynet thanks!
as your advice, I change the TypeDesc to `UINT8`.
`ImageSpec spec = ImageSpec(WIDTH, HEIGHT, CHANNELS, TypeDesc::UINT8);`
but,  I get the same wrong result.

**Comment by grdanny** [comment]
Those artifacts at the bottom of your image look like uninitialized memory.
I think this will do the trick:

```
    const int WIDTH = 640, HEIGHT = 480, CHANNELS = 3;
    unsigned char buffer[WIDTH * HEIGHT * CHANNELS];
    ImageSpec spec = ImageSpec(WIDTH, HEIGHT, CHANNELS);
    ImageBuf ib = ImageBuf(spec, buffer);
    ib.reset(spec, InitializePixels::Yes);
    ImageBufAlgo::render_text(ib, 0, 10, "Hello, world!");
    ib.write("oiio-render_text.png");
```

**Comment by nrusch** [comment]
> I believe you're having an issue with your buffer consisting of chars (8bit) and the default TypeDesc in oiio is FLOAT (32bit)

The default `TypeDesc` for a new `ImageSpec` is in fact actually `TypeUInt8` (see https://github.com/OpenImageIO/oiio/blob/dev-2.4/src/include/OpenImageIO/imageio.h#L331).

I think @grdanny is right that you're just seeing garbage pixels due to uninitialized memory. If you really want to wrap an external buffer (vs. just letting the OIIO API handle the memory), then I think it's simpler to just zero-initialize your buffer when you create it, e.g:

```c++
const int WIDTH = 640, HEIGHT = 480, CHANNELS = 3;
unsigned char buffer[WIDTH * HEIGHT * CHANNELS] = {0};  // Note the explicit initializer
ImageBuf ib(ImageSpec(WIDTH, HEIGHT, CHANNELS), buffer);
ImageBufAlgo::render_text(ib, 0, 10, "Hello, world!");
ib.write("oiio-render_text.png");
```

Now, the real issue here is that there is no text showing up in your result. Based on your example code, the text should be drawn in the upper-left corner of the image, and when I test this same code on Linux, I get text drawing in that corner as expected. Thus, it's possible that this is a platform-specific issue... I don't have a Mac, so I can't really dig any further in that direction.

Does it make a difference if you pass an explicit font name or even a full font file path to `ImageBufAlgo::render_text`?

**Comment by lgritz** [comment]
Good catch, @grdanny, on the uninitialized memory issue -- that certainly escaped my notice when I first read the issue description.

Another alternative to the `={0}` (does this really zero out the entire array and not just the first element? did that change recently?) is to explicitly call

    ImageBufAlgo::zero(ib);

As for the text, it's hard to tell from this what might be failing. Let's check the error:

    bool ok = ImageBufAlgo::render_text(ib, 0, 10, "Hello, world!");
    if (!ok)
        OIIO::print("render_text had an error: {}\n", ib.geterror());

Anybody want to place bets on what it will report? Did OIIO get built without freetype support? Was it unable to find any font at all? Something else?

**Comment by diuming** [comment]
@lgritz thanks
I didn't compile with FreeType. sorry about that.
error `render_text had an error: OpenImageIO was not compiled with FreeType for font rendering`

**Comment by lgritz** [comment]
Random idea for any onlookers to comment on: Does it make sense for the IB destructor to print any pending error messages that haven't been retrieved with geterror? This behavior could be disabled by a global option for advanced users who want to silence it, but having it on by default would prevent new users from failing to notice errors. 

That idea feels like it would be mildly annoying to experienced users (er, experienced users who intentionally ignore IBA errors?), but it might help a LOT of newcomers to OIIO who could then catch problems right away if the functions they are calling are producing error messages that they fail to notice. It would help a lot of people diagnose what's going wrong before they ever get frustrated and file an issue.

Any opinions?
